modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/opt/GridH2Table.java

```diff
@@ -1059,48 +1055,51 @@ public boolean rebuildFromHashInProgress() {
                 }
             }
 
-            if (registredIdx != null)
-                return registredIdx;
-        }
+            if (registredIdx != null) {
+                String idxCols = Stream.of(registredIdx.getIndexColumns())
+                    .map(k -> k.columnName).collect(Collectors.joining(", "));
 
-        return null;
+                U.warn(log, "Index with the given set or subset of columns already exists " +
+                    "(consider dropping either new or existing index) [cacheName=" + cacheInfo.name() + ", " +
+                    "schemaName=" + getSchema().getName() + ", tableName=" + getName() +
+                    ", newIndexName=" + curIdx.getName() + ", existingIndexName=" + registredIdx.getName() +
+                    ", existingIndexColumns=[" + idxCols + "]]");
+            }
+        }
     }
 
     /**
      * Add index that is in an intermediate state and is still being built, thus is not used in queries until it is
      * promoted.
      *
      * @param idx Index to add.
+     * @return {@code true} If index not intersects by name with an existing one.
      * @throws IgniteCheckedException If failed.
      */
-    public void proposeUserIndex(Index idx) throws IgniteCheckedException {
+    public boolean proposeUserIndex(Index idx) throws IgniteCheckedException {
         assert idx instanceof GridH2IndexBase;
 
         lock(true);
 
         try {
             ensureNotDestroyed();
 
-            Index idxExist = checkIndexPresence(idx);
-
-            if (idxExist != null) {
-                String idxCols = Stream.of(idxExist.getIndexColumns())
-                    .map(k -> k.columnName).collect(Collectors.joining(", "));
-
-                U.warn(log, "Index with the given set or subset of columns already exists " +
-                    "(consider dropping either new or existing index) [cacheName=" + cacheInfo.name() + ", " +
-                    "schemaName=" + getSchema().getName() + ", tableName=" + getName() +
-                    ", newIndexName=" + idx.getName() + ", existingIndexName=" + idxExist.getName() +
-                    ", existingIndexColumns=[" + idxCols + "]]");
+            for (Index idx0 : idxs) {
+                if (F.eq(idx.getName(), idx0.getName()))
+                    throw new SchemaOperationException(CODE_INDEX_EXISTS, idx.getName());
             }
 
+            checkEquivalentFieldsIndexIsPresent(idx);
+
             Index oldTmpIdx = tmpIdxs.put(idx.getName(), (GridH2IndexBase)idx);
 
             assert oldTmpIdx == null;
         }
         finally {
             unlock(true);
         }
+
+        return true;
```

Comment:
Looks like the method always returns true.

Reply:
got it